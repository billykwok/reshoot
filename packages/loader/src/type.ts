export type OutputShape = {
  mime: string;
  src: string | false;
  aspectRatio: string | false;
  srcSet: string | false;
  placeholder: string | false;
  color: string | false;
};

export enum AspectRatioType {
  WidthByHeight = 'widthByHeight',
  HeightByWidth = 'heightByWidth'
}

export enum AspectRatioFormat {
  Percent = 'percent',
  Ratio = 'ratio'
}

export type AspectRatio =
  | number
  | {
      type: AspectRatioType;
      format: AspectRatioFormat;
      decimal: number;
    };

export type Placeholder = {
  size: number;
  trimDataUrl: boolean;
};

export type InternalOutput = {
  mime: string;
  src?: string;
  aspectRatio?: number;
  srcSet?: string[];
  placeholder?: string;
  color?: string;
};

export type Output = {
  mime: string;
  src?: string;
  aspectRatio?: number;
  srcSet?: string;
  placeholder?: string;
  color?: string;
};

export type Options = {
  name: string;
  outputPath: string | ((path: string) => string) | null;
  publicPath: string | ((path: string) => string) | null;
  context: string;
  shape: OutputShape;
  srcSet: Array<number>;
  quality: number;
  background: string;
  color:
    | 'Vibrant'
    | 'Muted'
    | 'DarkVibrant'
    | 'DarkMuted'
    | 'LightVibrant'
    | 'LightMuted'
    | string;
  forceFormat: string | false;
  placeholder: Placeholder;
  aspectRatio: AspectRatio;
  disable: boolean;
  emitFile: boolean;
};
