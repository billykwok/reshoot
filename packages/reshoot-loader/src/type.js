// @flow
export type OutputShape = {
  src: string | false,
  aspectRatio: string | false,
  srcSet: string | false,
  placeholder: string | false,
  background: string | false,
  palette: string | false
};

export type AspectRatio = {
  type: 'widthByHeight' | 'heightByWidth',
  format: 'percent' | 'ratio',
  decimal: number
};

export type Placeholder = {
  size: number,
  trimDataUrl: boolean
};

export type Output = {
  src?: string,
  aspectRatio?: string,
  srcSet?: string,
  placeholder?: string,
  background?: string,
  palette?: string
};

export type Options = {
  name: string,
  outputPath: string | (() => string) | null,
  publicPath: string | (() => string) | null,
  context: string | (() => string) | null,
  shape: OutputShape,
  srcSet: Array<number>,
  quality: number,
  background: string,
  color:
    | 'Vibrant'
    | 'Muted'
    | 'DarkVibrant'
    | 'DarkMuted'
    | 'LightVibrant'
    | 'LightMuted'
    | string,
  forceFormat: string | false,
  placeholder: Placeholder,
  aspectRatio: AspectRatio,
  disable: boolean,
  emitFile: boolean
};
