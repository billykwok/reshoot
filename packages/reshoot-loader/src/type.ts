export type OutputShape = {
  mime: string;
  src: string | false;
  aspectRatio: string | false;
  srcSet: string | false;
  placeholder: string | false;
  color: string | false;
};

export type AspectRatio = {
  type: 'widthByHeight' | 'heightByWidth';
  format: 'percent' | 'ratio';
  decimal: number;
};

export type Placeholder = {
  size: number;
  trimDataUrl: boolean;
};

export type Output = {
  mime: string;
  src?: string;
  aspectRatio?: number;
  srcSet?: string;
  placeholder?: string;
  color?: string;
};

export type Options = {
  name: string;
  outputPath: string | ((path: string) => string) | null;
  publicPath: string | ((path: string) => string) | null;
  context: string | ((path: string) => string) | null;
  shape: OutputShape;
  srcSet: Array<number>;
  quality: number;
  background: string;
  color:
    | 'Vibrant'
    | 'Muted'
    | 'DarkVibrant'
    | 'DarkMuted'
    | 'LightVibrant'
    | 'LightMuted'
    | string;
  forceFormat: string | false;
  placeholder: Placeholder;
  aspectRatio: AspectRatio;
  disable: boolean;
  emitFile: boolean;
};
