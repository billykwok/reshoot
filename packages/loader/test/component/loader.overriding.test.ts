import { describe, beforeAll, afterAll, test, expect } from '@jest/globals';
import path from 'path';
import compile from './compiler';
import createMemfs from './createMemfs';
import { mockBase64, mockMetroHash } from './mock';

import type { ShapeArgument } from '../../src/type';

describe('Overriding component test for @reshoot/loader', () => {
  beforeAll(() => {
    jest.doMock('metrohash', mockMetroHash('tEsThAsH'));
    jest.doMock('../../src/util/base64', mockBase64('data/of=someimage'));
  });

  afterAll((done) => {
    jest.resetModules();
    setTimeout(() => ((done as unknown) as () => void)(), 1000).unref();
  });

  test('should handle custom shape successfully', async () => {
    const memfs = createMemfs();
    const actual = await compile(
      memfs,
      '../../../../__fixtures__/test-image.jpg',
      {
        outputPath: '/images/',
        cache: false,
        shape: ({ type, src, srcSet, sources, ...output }: ShapeArgument) => ({
          ...output,
          sources: [{ type, src, srcSet }, ...sources],
        }),
      }
    );
    const images = await memfs.promises.readdir(
      path.resolve(__dirname, './images')
    );
    expect(images).toEqual(
      expect.arrayContaining([
        'tEsThAsH-480.jpg',
        'tEsThAsH-640.jpg',
        'tEsThAsH-720.jpg',
        'tEsThAsH-840.jpg',
        'tEsThAsH-1080.jpg',
        'tEsThAsH-480.webp',
        'tEsThAsH-640.webp',
        'tEsThAsH-720.webp',
        'tEsThAsH-840.webp',
        'tEsThAsH-1080.webp',
      ])
    );
    expect(images).toHaveLength(10);
    expect(actual).toMatchSnapshot();
    memfs.reset();
  }, 30000);
});
