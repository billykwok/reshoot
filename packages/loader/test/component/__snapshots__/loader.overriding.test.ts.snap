// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overriding component test for @reshoot/loader should handle custom shape successfully 1`] = `"export default {\\"placeholder\\":\\"data:image/jpeg;base64,data/of=someimage\\",\\"width\\":4774,\\"height\\":3024,\\"aspectRatio\\":0.6334,\\"color\\":\\"#283838\\",\\"sources\\":[{\\"type\\":\\"image/jpeg\\",\\"src\\":__webpack_public_path__+\\"tEsThAsH-720.jpg\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-480.jpg 480w,\\"+__webpack_public_path__+\\"tEsThAsH-640.jpg 640w,\\"+__webpack_public_path__+\\"tEsThAsH-840.jpg 840w,\\"+__webpack_public_path__+\\"tEsThAsH-1080.jpg 1080w\\"},{\\"type\\":\\"image/webp\\",\\"src\\":__webpack_public_path__+\\"tEsThAsH-720.webp\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-480.webp 480w,\\"+__webpack_public_path__+\\"tEsThAsH-640.webp 640w,\\"+__webpack_public_path__+\\"tEsThAsH-840.webp 840w,\\"+__webpack_public_path__+\\"tEsThAsH-1080.webp 1080w\\"}]};"`;
